### PR TITLE
MONGOCRYPT-605 support named KMS providers

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary

This PR builds upon https://github.com/mongodb/libmongocrypt/pull/723 to implement support for named KMS providers.

The diff is large. I recommend reviewing individual commits.

# Background

Named KMS providers is further described in [DBX Scope: Support Named KMS Providers](https://docs.google.com/document/d/1IUxkgGFx5VLjMr7HWxBOXerN0pUZwcqWb11gLoRtcJs/edit?usp=sharing).

The KMS ID is a string identifier. Previously supported KMS IDs were only: "aws", "azure", "gcp", "kmip", and "local". The KMS ID is now expanded to support name suffixes. (e.g. "local:myname").

Named KMS providers enables more than one of each KMS provider type to be configured. Example: libmongocrypt can now support more than one "local" KMS provider configured:

```python
kms_providers = {
    "local": {"key": local_kek1},        # Unnamed KMS provider.
    "local:myname": {"key": local_kek2 } # Named KMS provider with name "myname".
}
```

## `csfle`

Minor changes have been made to the test utility `csfle` to improve the printed output for KMIP. This helped to generate test data. `csfle` may be used to dump KMIP request and responses with the `--trace` option. Example:

```bash
# Run this script to create a KMIP data key. Expects MongoDB server to be running on port 27017. Expects `kms_kmip_server.py` to be running from drivers-evergreen-tools
DRIVER_TOOLS=~/code/drivers-evergreen-tools
# KMS_PROVIDERS_FILE contains JSON for KMS providers credentials. Example: `{ "kmip": { "endpoint": "localhost:5698" } }`
KMS_PROVIDERS_FILE=~/.csfle/kms_providers.json
export KMIP_CLIENT_CERTIFICATE="$DRIVER_TOOLS/.evergreen/x509gen/client.pem"
export KMIP_CA_CERTIFICATE="$DRIVER_TOOLS/.evergreen/x509gen/ca.pem"
./cmake-build/csfle create_datakey \
    --kms_providers_file $KMS_PROVIDERS_FILE \
    --kms_provider kmip \
    --tls_ca_file $KMIP_CA_CERTIFICATE \
    --tls_certificate_key_file $KMIP_CLIENT_CERTIFICATE \
    --trace true
```

Results in output containing:
```
2024/01/05 14:38:18.0239: [89372]:    DEBUG:       mongoc: Current state = MONGOCRYPT_CTX_NEED_KMS
2024/01/05 14:38:18.0300: [89372]:    DEBUG:       mongoc: --> sending KMS message:
42007801000001504200770100000038420069010000002042006a0200000004000000010000000042006b0200000004000000000000000042000d0200000004000000010000000042000f010000010842005c0500000004000000030000000042007901000000f0420057050000000400000007000000004200910100000038420008010000003042000a070000001843727970746f67726170686963205573616765204d61736b42000b0200000004000000000000000042008501000000984200860500000004000000020000000042004001000000804200420500000004000000020000000042004501000000684200430800000060dfa14f0ec65c46e672dbe6d6cfd5df099b6d745b617f6cd23484d1256c68d89082e04032d9d73b362ea98bb3e1aaca7b0c36d6fdb9c5915732db2a855a1914a1170e88a5cc16f14902f63bde4b0f11ae2e3577169f5bd39b69f07892737274a3
2024/01/05 14:38:18.0300: [89372]:    DEBUG:       mongoc: <-- read KMS reply:
42007b010000009042007a0100000048420069010000002042006a0200000004000000010000000042006b0200000004000000000000000042009209000000080000000065985aaa42000d0200000004000000010000000042000f010000003842005c0500000004000000030000000042007f0500000004000000000000000042007c010000001042009407000000023433000000000000
```

The hex can be printed as XML with the [kmip_dump test utility](https://github.com/kevinAlbs/c-bootstrap/tree/cfa8e4224605b78ca252128325ad3281b33427ea/tools/kmip_dump)

### CODE REVIEW

File: src/mongocrypt-kms-ctx-private.h [hunk 1]
```diff
@@ -51,20 +51,23 @@ struct _mongocrypt_kms_ctx_t {
...
+    char *kmsid;
 };
 
 bool _mongocrypt_kms_ctx_init_aws_decrypt(mongocrypt_kms_ctx_t *kms,
                                           _mongocrypt_opts_kms_providers_t *kms_providers,
                                           _mongocrypt_key_doc_t *key,
                                           _mongocrypt_log_t *log,
-                                          _mongocrypt_crypto_t *crypto) MONGOCRYPT_WARN_UNUSED_RESULT;
+                                          _mongocrypt_crypto_t *crypto,
+                                          const char *kmsid) MONGOCRYPT_WARN_UNUSED_RESULT;
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mongocrypt-kms-ctx-private.h]: Updated to pass `log` as last argument.

File: test/util/util.c [hunk 2]
```diff
@@ -531,16 +531,32 @@ static bool _state_need_kms(_state_machine_t *state_machine, bson_error_t *error
...
-
         if (state_machine->trace) {
-            MONGOC_DEBUG("--> sending KMS message: \n%.*s", (int)iov.iov_len, (char *)iov.iov_base);
+            MONGOC_DEBUG("--> sending KMS message:");
+            const char *kms_provider = mongocrypt_kms_ctx_get_kms_provider(kms_ctx, NULL);
+            if (0 == strcmp(kms_provider, "kmip")) {
+                // Print KMIP protocol as hex.
+                uint8_t *as_u8_ptr = iov.iov_base;
+                for (size_t i = 0; i < iov.iov_len; i++) {
+                    printf("%02x", as_u8_ptr[i]);
```
Comment (vector-of-bool) [inline test/util/util.c:541]: Should this `printf` (and below) be a `MONGOC_DEBUG`?

**Review by vector-of-bool — APPROVED** [review]
LGTM. One minor comment.

File: test/util/util.c (hunk 2 above)
Comment (kevinAlbs) [reply test/util/util.c:541]: No. `printf` was intended. `MONGOC_DEBUG` adds newlines and a timestamp. This print is intended to dump hex for easy copy-pasting.

**Pull request merged**